judgels-backends/uriel/uriel-app/src/main/java/judgels/uriel/contest/scoreboard/icpc/IcpcScoreboardProcessor.java

```diff
@@ -50,4 +190,38 @@ public Scoreboard filterContestantJids(Scoreboard scoreboard, Set<String> contes
                 .content(new IcpcScoreboardContent.Builder().entries(filteredEntries).build())
                 .build();
     }
+
+    private List<IcpcScoreboardEntry> sortEntriesAndAssignRanks(List<IcpcScoreboardEntry> scoreboardEntries) {
+        Comparator<IcpcScoreboardEntry> comparator = new StandardIcpcScoreboardEntryComparator();
+        scoreboardEntries.sort(comparator);
+
+        List<IcpcScoreboardEntry> result = new ArrayList<>();
+
+        int currentRank = 1;
+        for (int i = 0; i < scoreboardEntries.size(); i++) {
+            if (i != 0 && comparator.compare(scoreboardEntries.get(i), scoreboardEntries.get(i - 1)) != 0) {
+                currentRank = i + 1;
+            }
+            result.add(new IcpcScoreboardEntry.Builder().from(scoreboardEntries.get(i)).rank(currentRank).build());
+        }
+
+        return result;
+    }
+
+    private Long submissionPenaltyInMillisecond(
+            Instant submissionTime,
+            Optional<Instant> contestantStartTime,
+            Instant contestStartTime) {
```

Comment:
```suggestion
            Instant contestBeginTime) {
```

(the semantics are different)